src/networkd.c

```diff
@@ -1460,29 +1474,41 @@ _netplan_netdef_write_networkd(
 gboolean
 _netplan_networkd_write_wait_online(const NetplanState* np_state, const char* rootdir)
 {
+    // Set of all current network interfaces, potentially non yet renamed
+    GHashTable* system_interfaces = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    _netplan_query_system_interfaces(system_interfaces);
```

Comment:
I'm not certain here about the impact of ignoring devices not currently present in /sys/class/net during systemd generator timeframe.  By only considering interfaces already present in /sys/class/net, are we exposed to scenarios where the device drivers for the specified device haven't loaded yet (in initramfs-less environments without the proper drivers compiled into the kernel)?   This reminds me of this cloud-init issue https://github.com/canonical/cloud-init/issues/4451 for minimal images.   In the event that we don't have device drivers loaded (from initramfs or kernel static compiled), is it possible netplan's generator may not see the device in /sys/class/net via `if_nameindex` and systemd-networkd-wait-online will be disabled right?  Maybe in this case, we prescribe that initramfs/kernel must have the desired device drivers in order to expect systemd-networkd-wait-online to block properly.

Reply:
A similar (the same?) issue will happen with virtual devices. By the time the generator runs they will not be created yet so they will not be in the list.

Reply:
James corrected me, this isn't systemd generator timeframe, but the point in time when `netplan  generate` is called.   For cloud-init init-local  boot stage, I think we may still be exposed to this race where cloud-init local gets in any invokes `netplan generate` possibly before supplemental device drivers have loaded. So this may still be a concern.

Reply:
Oh! That's a very interesting case.. I just pushed more commits to not require virtual devices (bridges, bonds, dummys, ...) to be available in `/sys/class/net` just yet. Those will be created later.

For physical devices that are not yet loaded, I don't know. I don't think there's anything that can be done as part of this PR and such interfaces might be ignored by `systemd-networkd-wait-online`.

A longer-term solution might be splitting up Netplan into a sytemd-generator and a systemd-service (running `Before=network-pre.target`, similar to `systemd-network-generator.service`), which we are tracking as spec "FO165 - Netplan generator architecture".

Do you think this should be a blocker?

Reply:
I think this retains the same gap/issue cloud-init and systemd-networkd-wait-online are already exposed to, so I don't think this needs to be fixed in this PR.  The stance we've taken already [Ubuntu minimal case](https://bugs.launchpad.net/cloud-images/+bug/2036968) was to recompile the kernel to include the module built-ins that are required to support virtio-net to ensure they are loaded in time for network generation and network-online for "initrdless" boots. It feels "ok" to state that if you have a critical device driver needed in early boot for your cloud image and you are booting without initramrd, you would either need to statically compile the necessary device drivers into that kernel, or use initrd and provide any supplemental drivers needed earlier boot. This may just be something we take into consideration for future behavior though, and thinking of how to approach this type of concern later in F0165 probably makes the most sense at the moment.